File: ReaderApp/launch.js
#!/usr/bin/env node
/**
 * launch.js — Start the Heaven's River audiobook reader in your browser.
 *
 * Usage:  node launch.js
 *         node launch.js --port 8080
 */

const http = require("http");
const fs   = require("fs");
const path = require("path");

const args = process.argv.slice(2);
const portArg = args.indexOf("--port");
const parsedPort = portArg !== -1 ? parseInt(args[portArg + 1], 10) : parseInt(args[0], 10);
const PORT = Number.isFinite(parsedPort) && parsedPort > 0 ? parsedPort : 8765;
const ROOT = path.resolve(__dirname, "..");
const URL  = `http://localhost:${PORT}/ReaderApp/reader.html`;

const MIME = {
  ".html": "text/html",
  ".js":   "application/javascript",
  ".css":  "text/css",
  ".json": "application/json",
  ".mp3":  "audio/mpeg",
  ".png":  "image/png",
  ".ico":  "image/x-icon",
};

const server = http.createServer((req, res) => {
  // Strip query string, decode URI, prevent path traversal
  let urlPath = decodeURIComponent(req.url.split("?")[0]);
  if (urlPath === "/") urlPath = "/ReaderApp/reader.html";
  const filePath = path.join(ROOT, urlPath);

  // Block traversal outside ROOT
  if (!filePath.startsWith(ROOT + path.sep) && filePath !== ROOT) {
    res.writeHead(403);
    res.end("Forbidden");
    return;
  }

  fs.stat(filePath, (err, stat) => {
    if (err || !stat.isFile()) {
      res.writeHead(404);
      res.end("Not found");
      return;
    }
    const ext  = path.extname(filePath).toLowerCase();
    const mime = MIME[ext] || "application/octet-stream";
    res.writeHead(200, { "Content-Type": mime, "Content-Length": stat.size });
    fs.createReadStream(filePath).pipe(res);
  });
});

server.listen(PORT, "127.0.0.1", () => {
  console.log("");
  console.log("  Heaven's River Reader");
  console.log("  ───────────────────────────────");
  console.log(`  Serving from : ${ROOT}`);
  console.log(`  Open in browser: ${URL}`);
  console.log("");
  console.log("  Press Ctrl+C to stop.");
  console.log("");

  // Open browser after server is listening
  const { execSync } = require("child_process");
  const open =
    process.platform === "win32"  ? `start "" "${URL}"` :
    process.platform === "darwin" ? `open "${URL}"` :
                                    `xdg-open "${URL}"`;
  try { execSync(open); } catch (_) {}
});

process.on("SIGINT", () => {
  console.log("\n  Server stopped.");
  process.exit(0);
});

File: ReaderApp/launch.py
#!/usr/bin/env python3
"""
launch.py — Start the Heaven's River audiobook reader.

Usage:  python3 launch.py
        python3 launch.py --port 8080
"""
import argparse
import os
import sys
import webbrowser
from http.server import HTTPServer, SimpleHTTPRequestHandler
from threading import Timer

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--port', type=int, default=8765)
    args = parser.parse_args()

    port = args.port
    url  = f'http://localhost:{port}/reader.html'

    # Change to the directory containing this script
    script_dir = os.path.dirname(os.path.abspath(__file__))
    os.chdir(script_dir)

    # Open browser after a short delay (give server time to start)
    Timer(0.8, lambda: webbrowser.open(url)).start()

    print(f'')
    print(f'  📚  Heaven\'s River Reader')
    print(f'  ──────────────────────────')
    print(f'  Serving from : {script_dir}')
    print(f'  Open in browser: {url}')
    print(f'')
    print(f'  Press Ctrl+C to stop.')
    print(f'')

    class QuietHandler(SimpleHTTPRequestHandler):
        def log_message(self, fmt, *args):
            pass  # silence request logs

    server = HTTPServer(('localhost', port), QuietHandler)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print('\n  Server stopped.')

if __name__ == '__main__':
    main()
